File: Algorithms_python/105_Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        if not preorder:
            return None
        v = preorder[0]
        root = TreeNode(val = v)
        i = inorder.index(v)
        root.left = self.buildTree(preorder[1: 1+i], inorder[:i])
        root.right = self.buildTree(preorder[1+i: ], inorder[i + 1: ])
        return root

'''
前序遍历：先遍历根节点，再遍历左右子树；中序遍历：先遍历左子树，再遍历根节点，最后遍历右子树。
这题思路是：前序序列的第一个结点做为根节点，然后在中序序列中找到根节点的位置 i，
再将中序序列划分为左子树 inorder[:i] 、右子树 inorder[i+1:]，以此来构造整个树。
'''

File: Algorithms_C++/105_Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    unordered_map<int, int> d;
    TreeNode* dfs(vector<int>& preorder, vector<int>& inorder, int i, int j, int n){
        if(n <= 0){
            return nullptr;
        }
        int v = preorder[i];
        int k = d[v];
        TreeNode* l = dfs(preorder, inorder, i + 1, j, k - j);
        TreeNode* r = dfs(preorder, inorder, i + 1 + k - j, k + 1, n - 1 - (k - j));
        return new TreeNode(v, l, r);
    }
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        int n = preorder.size();
        for(int i = 0; i < n; ++i){
            d[inorder[i]] = i;
        }
        return dfs(preorder, inorder, 0, 0, n);
    }
};